solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java

```diff
@@ -146,5 +201,137 @@ public float getMaxScore(int upTo) throws IOException {
         return Float.POSITIVE_INFINITY;
       }
     }
+
+    /**
+     * A FeatureScorer that reads the numeric docValues for a field
+     */
+    public final class NumericDocValuesFieldValueFeatureScorer extends FeatureScorer {
+      private final NumericDocValues docValues;
+      private final NumberType numberType;
+
+      public NumericDocValuesFieldValueFeatureScorer(final FeatureWeight weight, final LeafReaderContext context,
+                                              final DocIdSetIterator itr, final NumberType numberType) {
+        super(weight, itr);
+        this.numberType = numberType;
+
+        NumericDocValues docValues;
+        try {
+          docValues = DocValues.getNumeric(context.reader(), field);
+        } catch (IOException e) {
+          throw new IllegalArgumentException("Could not read numeric docValues for field " + field);
+        }
+        this.docValues = docValues;
+      }
+
+      @Override
+      public float score() throws IOException {
+        if (docValues.advanceExact(itr.docID())) {
+          return readNumericDocValues();
+        }
+        return FieldValueFeature.this.getDefaultValue();
+      }
+
+      /**
+       * Read the numeric value for a field and convert the different number types to float.
+       *
+       * @return The numeric value that the docValues contain for the current document
+       * @throws IOException if docValues cannot be read
+       */
+      private float readNumericDocValues() throws IOException {
+        if (NumberType.FLOAT.equals(numberType)) {
+          // convert float value that was stored as long back to float
+          return Float.intBitsToFloat((int) docValues.longValue());
+        } else if (NumberType.DOUBLE.equals(numberType)) {
+          // handle double value conversion
+          return (float) Double.longBitsToDouble(docValues.longValue());
+        }
+        // just take the long value
+        return docValues.longValue();
+      }
```

Comment:
1/2 So I was trying to learn and understand better w.r.t. why the `intBitsToFloat` and `longBitsToDouble` are the conversions to use here and that led to https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java#L603-L644 and the realisation that `DATE` is a `numberType` possibility!

Would you have any thoughts on using a `switch` statement here with FLOAT/DOUBLE/INT/LONG all valid and to throw an exception for anything else ...

Reply:
Somewhat strangely so perhaps the use of `DATE` fields "just works" for both the existing stored and the new docValues implementations. I've added a `TestFieldValueFeature.testThatDateValuesAreCorrectlyParsed` test to demonstrate that and so then no need to change the implementation here.

Reply:
Hm... the date being a numeric type really is quite unintuitive.
I am not really fond of copying the switch-logic that you referenced, but sadly the method is private. :/

I will go and make a tea and then look into that.
Thank you for adding the test for the date field. The new clarity on what value is returned for the date type is very good.

Reply:
Oh... Thanks github for not refreshing the page... :roll_eyes: 
After looking into that I saw the same behavior as you did and was just about to comment that. :D

Reply:
> ... I am not really fond of copying the switch-logic that you referenced, but sadly the method is private. ...

Likewise, I would have preferred for there to be some existing reusable method and was surprised that there isn't one (or not an obviously findable one). The `SolrDocumentFetcher` one has the `sortableNumeric` flag (which we don't need here) and some special logic for `LatLonPointSpatialField` and `AbstractEnumField` scenarios (also not applicable here) ... otherwise it would have been a clear _"okay, let's factor out a method here and put it somewhere that both SolrDocumentFeature and FieldValueFeature can use it"_ kind of scenario. Oh well.